### PR TITLE
internal/docstore/firedocstore: add URL opener

### PR DESCRIPTION
**Pull request by shantuo**
Updates #1651

### CODE REVIEW

File: internal/docstore/firedocstore/fs.go [hunk 1]
```diff
@@ -12,26 +12,150 @@
...
 // limitations under the License.
 
+// Package firedocstore provides a docstore implementation backed by GCP
+// Firestore.
+// Use OpenCollection to construct a *docstore.Collection.
+//
+// URLs
+//
+// For docstore.OpenCollection, firedocstore registers for the scheme
+// "firedocstore".
```
Comment (vangent) [inline internal/docstore/firedocstore/fs.go]: Same as other PRs: "firestore" is probably fine.

File: internal/docstore/firedocstore/fs.go [hunk 2]
```diff
@@ -12,26 +12,150 @@
...
+// The default URL opener will create a connection using use default
```
Comment (vangent) [inline internal/docstore/firedocstore/fs.go]: Nit: "using use" -> "using"

File: internal/docstore/firedocstore/fs.go [hunk 3]
```diff
@@ -12,26 +12,150 @@
...
+		return nil, fmt.Errorf("open collection %s: invalid query parameter %q", u, param)
+	}
+	cn, err := collectionNameFromURL(u)
+	if err != nil {
+		return nil, fmt.Errorf("open collection %s: %v", u, err)
+	}
+	return OpenCollection(o.Client, cn.ProjectID, cn.CollPath, nameField), nil
+}
+
+// CollectionName identifies a collection name.
```
Comment (vangent) [inline internal/docstore/firedocstore/fs.go]: I would leave this out. It's exported, but not usable anywhere. Just parse the individual fields in the `URLOpener` as we do elsewhere.

#1860 discusses making compound ID handling more consistent across the board.

File: internal/docstore/firedocstore/fs.go (hunk 1 above)
Comment (shantuo) [reply internal/docstore/firedocstore/fs.go]: Done.

File: internal/docstore/firedocstore/fs.go (hunk 2 above)
Comment (shantuo) [reply internal/docstore/firedocstore/fs.go]: Done.

File: internal/docstore/firedocstore/fs.go (hunk 3 above)
Comment (shantuo) [reply internal/docstore/firedocstore/fs.go]: Done.

File: internal/docstore/firedocstore/fs.go [hunk 4]
```diff
@@ -12,26 +12,136 @@
...
+// Dial returns a client to use with Firestore and a clean-up function to close
+// the client after used.
+func Dial(ctx context.Context, ts gcp.TokenSource) (*vkit.Client, func(), error) {
+	c, err := vkit.NewClient(ctx, option.WithTokenSource(ts), useragent.ClientOption("docstore"))
+	return c, func() { c.Close() }, err
+}
+
+// Scheme is the URL scheme firedocstore registers its URLOpener under on
+// docstore.DefaultMux.
+const Scheme = "firedocstore"
```
Comment (vangent) [inline internal/docstore/firedocstore/fs.go]: firedocstore -> firestore

File: internal/docstore/firedocstore/fs_test.go [hunk 5]
```diff
@@ -104,3 +108,64 @@ func TestConformance(t *testing.T) {
...
+
+func TestCollectionNameFromURL(t *testing.T) {
+	tests := []struct {
+		URL          string
+		WantErr      bool
+		WantProject  string
+		WantCollPath string
+	}{
+		{"firedocstore://proj/coll", false, "proj", "coll"},
+		{"firedocstore://proj/coll/doc/subcoll", false, "proj", "coll/doc/subcoll"},
```
Comment (vangent) [inline internal/docstore/firedocstore/fs_test.go]: firestore here too.

File: internal/docstore/firedocstore/fs_test.go [hunk 6]
```diff
@@ -104,3 +108,64 @@ func TestConformance(t *testing.T) {
...
+	tests := []struct {
+		URL     string
+		WantErr bool
+	}{
+		// OK.
+		{"firedocstore://myproject/mycoll?name_field=_id", false},
+		// OK, hierarchical collection.
+		{"firedocstore://myproject/mycoll/mydoc/subcoll?name_field=_id", false},
+		// Missing project ID.
+		{"firedocstore:///mycoll?name_field=_id", true},
```
Comment (vangent) [inline internal/docstore/firedocstore/fs_test.go]: and here

Comment (shantuo) [reply internal/docstore/firedocstore/fs_test.go]: Done.

**Pull request merged**